### PR TITLE
Linear decoders implementation + regression models + test case + examples

### PR DESCRIPTION
**Pull request by ChangLiu2022**

### CODE REVIEW

File: examples/decoder/kalman_filter.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+import yaml
+from sklearn.metrics import r2_score
+from tqdm import tqdm
+import matplotlib.pyplot as plt
+from neuraldecoding.utils import parse_verify_config
+from hydra import initialize, compose
+
+# Example script of decoding data using ridge regression with decoder
+# Data path needs to be specified first here:
+data_path = "/mnt/D8C4D588C4D56970/ND/github/LINK_dataset/data/pickles"
```
Comment (lhcubillos) [inline examples/decoder/kalman_filter.py:18]: Don't do the example with the pickle files, but rather with NWB files

File: examples/decoder/kalman_filter.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+# Train-test split
+test_len = np.min((len(TrialIndex)-1, 399))
+neural_train= sbp[:TrialIndex[300]]
+neural_test = sbp[TrialIndex[300]:TrialIndex[test_len]]
+finger_train = beh[:TrialIndex[300]]
+finger_test = beh[TrialIndex[300]:TrialIndex[test_len]]
+
+# Add history
+def add_hist(X, Y, hist=10):
```
Comment (lhcubillos) [inline examples/decoder/kalman_filter.py:54]: This should go in some feature extraction related file and should just be imported here

File: examples/decoder/kalman_filter.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    if j == 500 or j == 1000 or j == 1500:
+        # Example intervention
+        print(f"Intervening initialization at index {j}")
+        new_yhat = np.expand_dims(np.array([ 7.6127e-01,  4.8506e-01, -2.6989e-04, -5.7489e-05]),axis=0)
+        decoder.model.initialize(new_yhat)
+
+    prediction = decoder.predict(neural_test[j,:].unsqueeze(0))
+    predictions_online[j, :] = prediction
+
+corr_offline, _ = pearsonr(predictions_offline, finger_test)
```
Comment (lhcubillos) [inline examples/decoder/kalman_filter.py:101]: From this point on it should call functions defined in some evaluation related file

File: examples/decoder/lstm.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    """Loads the synthetic train and validation data from .npy files."""
+    data_train = np.load(os.path.join("tests", "trainer", "data", "train.npz"), allow_pickle=True)
+    train_X, train_Y = data_train["X"], data_train["Y"]
+    
+    data_valid = np.load(os.path.join("tests", "trainer", "data", "valid.npz"), allow_pickle=True)
+    valid_X, valid_Y = data_valid["X"], data_valid["Y"]
+    
+    return train_X, train_Y, valid_X, valid_Y
+
+def create_dataloaders(train_X, train_Y, valid_X, valid_Y, batch_size=32):
```
Comment (lhcubillos) [inline examples/decoder/lstm.py:32]: This should be defined as a util somewhere

File: examples/decoder/models/ridge_regression_model.pkl
Comment (lhcubillos) [inline examples/decoder/models/ridge_regression_model.pkl:1]: Make sure to not upload the .pkl files. Possibly add a .pkl ignore to .gitignore

File: examples/decoder/lstm.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+from neuraldecoding.utils import parse_verify_config
+from hydra import initialize, compose
+
+#prep
+cfg_path = os.path.join("..","..","configs","example_LSTM")
+
+with initialize(version_base=None, config_path=cfg_path):
+    config = compose("config")
+
+def load_synthetic_data():
```
Comment (lhcubillos) [inline examples/decoder/lstm.py:22]: This should also be imported as a util from somewhere

File: README.md [hunk 6]
```diff
@@ -1 +1,17 @@
-# neural_decoding
\ No newline at end of file
+# neural_decoding
+
+1. Add conda forge to conda channel: `conda config --append channels conda-forge`
+
+2. Install environment: `conda create -n ND —file requirement.txt python=3.12`
```
Comment (lhcubillos) [inline README.md:5]: replace the combined -- before file, to two dashes in a row "--". It fails otherwise

File: README.md [hunk 7]
```diff
@@ -1 +1,17 @@
...
+
+3. Install pytorch from: `https://pytorch.org/get-started/locally/`
```
Comment (lhcubillos) [inline README.md:7]: Add this to requirements.txt rather than having it here

**Review by lhcubillos — CHANGES_REQUESTED** [review]
Look through some of these comments. We can meet sometime next week to address them

File: neuraldecoding/trainer/NeuralNetworkTrainer.py [hunk 8]
```diff
@@ -1,191 +1,130 @@
...
-        """Creates and returns a learning rate scheduler based on the configuration."""
-        scheduler_class = getattr(torch.optim.lr_scheduler, scheduler_config.type)
-        return scheduler_class(optimizer, **scheduler_config.params)
-
-    def create_loss_function(self, loss_config: DictConfig) -> torch.nn.Module:
-        """Creates and returns a loss function based on the configuration."""
-        loss_class = getattr(torch.nn, loss_config.type)
-        return loss_class(**loss_config.params)
-    
-    def calc_corr(self, y1,y2):
```
Comment (htemmar) [inline neuraldecoding/trainer/NeuralNetworkTrainer.py:56]: move to utils, make eval part of config

**Comment by htemmar** [comment]
remove train one epoch from nn models, only have train_step - make default train step in top level nn model class which: 

train_step(x,y,loss_func,opt):
    yh = predict(x)
    loss = loss_func(y,yh)
    loss.backward()
    opt.step()
    return loss
